File: lib/cpp-wrappers/crop.js
const { throwError, isNodePattern } = require('@jimp/utils');
const Jimp = require('jimp');
const nativeAddon = require('../addon');
const { ensureInteger, isSynchronous, cppCallbackWrapper } = require('../util');

module.exports = {
    crop (x, y, w, h, cb) {
        try {
            x = Math.max(ensureInteger(x), 0);
            y = Math.max(ensureInteger(y), 0);
            w = Math.min(ensureInteger(w), this.getWidth());
            h = Math.min(ensureInteger(h), this.getHeight());

            const updateBitmap = () => {
                this.bitmap.data = this.bitmap.data.slice(0, w * h * 4);
                this.bitmap.width = w;
                this.bitmap.height = h;
            };

            const skip = (x === 0 && y === 0 && w === this.getWidth());
        
            // If X and Y is zero and w is the full image width then we can just cut the buffer to size.
            if (!skip) {
                nativeAddon.crop(
                    this.bitmap.data,
                    this.getWidth(),
                    this.getHeight(),
                    x,
                    y,
                    w,
                    h,
                    cppCallbackWrapper(this, typeof cb === 'function' ? (err) => {
                        if (err) {
                            return cb(err);
                        }

                        updateBitmap();

                        cb(null, this);
                    } : null)
                );
            } 
        
            if (isSynchronous(this, cb) || skip) {
                updateBitmap();
            }
        
            if(isNodePattern(cb) && (isSynchronous(this, cb) || skip)) {
                cb.call(this, null, this);
            }
        } catch (err) {
            return throwError.call(this, err, cb);
        }
        
        return this;
    },
    
    autocrop(...args) {
        let cb;
        let leaveBorder = 0;
        let tolerance = 0.0002;
        let cropOnlyFrames = true;
        let symmetric = false;

        let north = true;
        let east = true;
        let south = true;
        let west = true;

        for (const arg of args) {
            if (typeof arg === 'boolean') {
                cropOnlyFrames = arg;
            }

            if (typeof arg === 'number') {
                tolerance = arg;
            }

            if (typeof arg === 'function') {
                cb = arg;
            }

            if (typeof arg === 'object') {
                if (typeof arg.tolerance === 'number') {
                    tolerance = arg.tolerance;
                }

                if (typeof arg.cropOnlyFrames === 'boolean') {
                    cropOnlyFrames = arg.cropOnlyFrames;
                }

                if (typeof arg.cropSymmetric === 'boolean') {
                    symmetric = arg.cropSymmetric;
                }

                if (typeof arg.leaveBorder === 'number') {
                    leaveBorder = arg.leaveBorder;
                }

                if (typeof arg.ignoreSides === 'object') {
                    north = !arg.ignoreSides.north;
                    east = !arg.ignoreSides.east;
                    south = !arg.ignoreSides.south;
                    west = !arg.ignoreSides.west;
                }
            }
        }

        try {
            nativeAddon.autocrop(
                this.bitmap.data,
                this.getWidth(),
                this.getHeight(),
                ensureInteger(leaveBorder),
                tolerance,
                cropOnlyFrames,
                symmetric,
                north,
                east,
                south,
                west,
                this.bitmap,
                cppCallbackWrapper(this, typeof cb == 'function' ? (err, newWidth, newHeight, jimpInstance) => {
                    this.bitmap.width = newWidth;
                    this.bitmap.height = newHeight;
                    this.bitmap.data = this.bitmap.data.slice(0, this.bitmap.width * this.bitmap.height * 4);
                    cb(err, jimpInstance);
                }: null)
            );

            if (isSynchronous(this, cb)) {
                this.bitmap.data = this.bitmap.data.slice(0, this.bitmap.width * this.bitmap.height * 4);
            }
        } catch (err) {
            return throwError.call(this, err, cb);
        }

        if (isNodePattern(cb) && isSynchronous(this, cb)) {
            cb.call(this, null, this);
        }

        return this;
    }
};

File: cppsrc/modules/crop.cpp
#include "crop.hpp"
#include "cmath"
#include "../util/colour.hpp"

void crop (Image& image, long xOffset, long yOffset, long width, long height) {
    width = std::max(std::min(width, image.width), (long) 0);
    height = std::max(std::min(height, image.height), (long) 0);

    xOffset = std::max(xOffset, (long) 0);
    yOffset = std::max(yOffset, (long) 0);

    if (xOffset + width > image.width) {
        width = image.width - (image.width - width);
    }
    
    if (yOffset + height > image.height) {
        height = image.height - (image.height - height);
    }

    if (width == 0 || height == 0) {
        return;
    }

    uint8_t* imageData = image.getPixels();
    long offset = 0;

    for (long y = 0; height > y; y++) {
        for (long x = 0; width > x; x++) {
            uint8_t* samplePixel = image.getPixelAt(x + xOffset, y + yOffset);

            if (samplePixel == nullptr) {
                continue;
            }

            imageData[offset++] = samplePixel[0];
            imageData[offset++] = samplePixel[1];
            imageData[offset++] = samplePixel[2];
            imageData[offset++] = samplePixel[3];
        }
    }
}

void autocrop (Image& image, long leaveBorder, double tolerance, bool cropOnlyFrames, bool symmetric, bool north, bool east, bool south, bool west, long& newWidth, long& newHeight) {
    long northPixels = 0;
    long eastPixels = 0;
    long southPixels = 0;
    long westPixels = 0;

    if (cropOnlyFrames && (!north || !east || !south || !west)) {
        return;
    }

    scanLoop:
    if (north) {
        uint8_t* targetColour = image.getPixelAt(0, 0);

        for (long y = 0; image.height > y; y++) {
            for (long x = 0; image.width > x; x++) {
                uint8_t* sampleColour = image.getPixelAt(x, y);

                if (diff(sampleColour, targetColour) > tolerance) {
                    north = false;
                    goto scanLoop;
                }
            }

            northPixels++;
        }
    }

    if (east) {
        uint8_t* targetColour = image.getPixelAt(image.width - 1, 0);

        for (long x = image.width - 1; x >= 0; x--) {
            for (long y = 0; image.height > y; y++) {
                uint8_t* sampleColour = image.getPixelAt(x, y);

                if (diff(sampleColour, targetColour) > tolerance) {
                    east = false;
                    goto scanLoop;
                }
            }

            eastPixels++;
        }
    }

    if (south) {
        uint8_t* targetColour = image.getPixelAt(image.width - 1, image.height - 1);

        for (long y = image.height - 1; y >= 0; y--) {
            for (long x = 0; image.width > x; x++) {
                uint8_t* sampleColour = image.getPixelAt(x, y);

                if (diff(sampleColour, targetColour) > tolerance) {
                    south = false;
                    goto scanLoop;
                }
            }

            southPixels++;
        }
    }

    if (west) {
        uint8_t* targetColour = image.getPixelAt(0, image.height - 1);

        for (long x = 0; image.width > x; x++) {
            for (long y = 0; image.height > y; y++) {
                uint8_t* sampleColour = image.getPixelAt(x, y);

                if (diff(sampleColour, targetColour) > tolerance) {
                    west = false;
                    goto scanLoop;
                }
            }
            
            westPixels++;
        }
    }

    northPixels -= leaveBorder;
    eastPixels -= leaveBorder;
    southPixels -= leaveBorder;
    westPixels -= leaveBorder;

    if (cropOnlyFrames && (northPixels < 1 || eastPixels < 1 || southPixels < 1 || westPixels < 1)) {
        return;
    }

    if (symmetric) {
        northPixels = std::min(northPixels, southPixels);
        southPixels = northPixels;
        westPixels = std::min(westPixels, eastPixels);
        eastPixels = westPixels;
    }

    long xOffset = std::max(westPixels, 0l);
    long yOffset = std::max(northPixels, 0l);
    long width = std::max(image.width - eastPixels - xOffset, 1l);
    long height = std::max(image.height - southPixels - yOffset, 1l);

    crop(image, xOffset, yOffset, width, height);

    newWidth = width;
    newHeight = height;
}
